### PR TITLE
Feature/rename keys boto2 utils

### PR DESCRIPTION
**Pull request by ivche**
Add a function that renames keys on s3 (boto3) - make use of a modification function and a filter

### CODE REVIEW

File: aws_utils/s3/s3_utils.py [hunk 1]
```diff
@@ -410,3 +411,53 @@ def get_contents_of_directory(directory, bucket=None):
...
+    Rename a single file on S3 (=> copy + delete)
+    Args:
+        boto3_client (botocore.client.S3) : boto3 client
+        bucket_name (str):
+        old_prefix (str):
+        new_prefix (str):
+    """
+    if old_prefix != new_prefix:
+        if not bucket_name:
+            raise (ClientError, AttributeError)
```
Comment (DanielSlaterSK) [inline aws_utils/s3/s3_utils.py]: This should probably be a value error.

File: aws_utils/s3/s3_utils.py [hunk 2]
```diff
@@ -410,3 +411,53 @@ def get_contents_of_directory(directory, bucket=None):
...
+                      filter_keys_func=None):
+    """
+    Renames all keys in a prefix_root/folder/ which are not filtered by a filter function
+    Args:
+        bucket_name (str): An existing s3 bucket
+        bucket_region (str): A valid s3 region e.g 'us-east-1'
+        prefix_root (str): The prefix to start the recursive renaming with
+        prefix_modification_func (function): A function that changes the prefix string. In case it is None, nothing will be moved
+        filter_keys_func (function): A function that will apply a filter to the keys we don't want to move. In case it is None, nothing will be moved
+    """
```
Comment (DanielSlaterSK) [inline aws_utils/s3/s3_utils.py]: Would be really good to give an example here, use the googledocs format for the example is

Example:

**Comment by DanielSlaterSK** [comment]
:+1:

File: aws_utils/s3/s3_utils.py [hunk 3]
```diff
@@ -442,6 +442,11 @@ def rename_keys_on_s3(bucket_name, bucket_region, prefix_root, prefix_modificati
         prefix_root (str): The prefix to start the recursive renaming with
         prefix_modification_func (function): A function that changes the prefix string. In case it is None, nothing will be moved
         filter_keys_func (function): A function that will apply a filter to the keys we don't want to move. In case it is None, nothing will be moved
+    Example:
+        The function will expect having all the arguments without a default value :
+        def foo(s) -> str: return s.replace('a', 'b')
+        def bar(s) -> bool: return '.gz' in s
+        rename_keys_on_s3('audience', 'us-east-1', 'production/dsp/partner_name/', prefix_modification_func=foo, filter_keys_func=bar)
     """
     if prefix_modification_func is None:
```
Comment (DanielSlaterSK) [inline aws_utils/s3/s3_utils.py]: if we are going to put in an exception like this, lets do it the proper way, should be a TypeError be specify what the correct type is. The check should be for isinstance(prefix_modification_func, Callable)

**Comment by ivche** [comment]
The tox tests are failing on circleci with a timeout (which I assume is a circleci issue).
However all the tox tests pass now when we run `tox` locally

**Pull request merged**